src/surface.cpp

```diff
@@ -0,0 +1,1239 @@
+  if (n_words != 4) {
+    std::string err_msg{"Surface "};
+    err_msg += std::to_string(surf_id);
+    err_msg += " expects 4 coeffs but was given ";
+    err_msg += std::to_string(n_words);
+    fatal_error(err_msg);
+  }
+
+  // Parse the coefficients.
+  int stat = sscanf(coeffs.c_str(), "%lf %lf %lf %lf", &c1, &c2, &c3, &c4);
+  if (stat != 4) {
+    fatal_error("Something went wrong reading surface coeffs");
+  }
+}
+
+void read_coeffs(pugi::xml_node surf_node, int surf_id, double &c1, double &c2,
+                 double &c3, double &c4, double &c5, double &c6, double &c7,
+                 double &c8, double &c9, double &c10)
+{
+  // Check the given number of coefficients.
+  std::string coeffs = get_node_value(surf_node, "coeffs");
+  int n_words = word_count(coeffs);
+  if (n_words != 10) {
+    std::string err_msg{"Surface "};
+    err_msg += std::to_string(surf_id);
+    err_msg += " expects 10 coeffs but was given ";
+    err_msg += std::to_string(n_words);
+    fatal_error(err_msg);
+  }
+
+  // Parse the coefficients.
+  int stat = sscanf(coeffs.c_str(), "%lf %lf %lf %lf %lf %lf %lf %lf %lf %lf",
+                    &c1, &c2, &c3, &c4, &c5, &c6, &c7, &c8, &c9, &c10);
+  if (stat != 10) {
+    fatal_error("Something went wrong reading surface coeffs");
+  }
+}
+
+//==============================================================================
+// Surface implementation
+//==============================================================================
+
+Surface::Surface(pugi::xml_node surf_node)
+{
+  if (check_for_node(surf_node, "id")) {
+    id = stoi(get_node_value(surf_node, "id"));
+  } else {
+    fatal_error("Must specify id of surface in geometry XML file.");
+  }
+
+  if (check_for_node(surf_node, "name")) {
+    name = get_node_value(surf_node, "name");
+  }
+
+  if (check_for_node(surf_node, "boundary")) {
+    std::string surf_bc = get_node_value(surf_node, "boundary");
+
+    if (surf_bc.compare("transmission") == 0
+         or surf_bc.compare("transmit") == 0
+         or surf_bc.compare("") == 0) {
+      bc = BC_TRANSMIT;
+
+    } else if (surf_bc.compare("vacuum") == 0) {
+      bc = BC_VACUUM;
+
+    } else if (surf_bc.compare("reflective") == 0
+               or surf_bc.compare("reflect") == 0
+               or surf_bc.compare("reflecting") == 0) {
+      bc = BC_REFLECT;
+    } else if (surf_bc.compare("periodic") == 0) {
+      bc = BC_PERIODIC;
+    } else {
+      std::string err_msg("Unknown boundary condition \"");
+      err_msg += surf_bc + "\" specified on surface " + std::to_string(id);
+      fatal_error(err_msg);
+    }
+
+  } else {
+    bc = BC_TRANSMIT;
+  }
+
+}
+
+bool
+Surface::sense(const double xyz[3], const double uvw[3]) const
+{
+  // Evaluate the surface equation at the particle's coordinates to determine
+  // which side the particle is on.
+  const double f = evaluate(xyz);
+
+  // Check which side of surface the point is on.
+  if (fabs(f) < FP_COINCIDENT) {
```

Comment:
C++ actually provides floating-point overloads for `abs`, so I would just use that instead of the C-like `fabs`